### PR TITLE
eth: enforce announcement metadatas and drop peers violating the protocol

### PR DESCRIPTION
**Pull request by karalabe**
In `eth/68` we introduced the capability to announce transaction types and sizes along with their hashes. The goal was to allow fetching large transactions and blob transactions via a throttled mechanism to avoid saturating a connection. Whilst we did start announcing the extra data, we never started validating it. This PR adds that component, cross checking delivered transaction bodies with announced metadatas and dropping offending peers. This is a **necessary** DoS protection, otherwise peers could announce fake small blob txs and cause us to request many concurrently, overloading our bandwidth.

### CODE REVIEW

File: tests/fuzzers/txfetcher/txfetcher_fuzzer.go [hunk 1]
```diff
@@ -128,7 +129,7 @@ func Fuzz(input []byte) int {
 			if verbose {
 				fmt.Println("Notify", peer, announceIdxs)
 			}
-			if err := f.Notify(peer, announces); err != nil {
```
Comment (MariusVanDerWijden) [inline tests/fuzzers/txfetcher/txfetcher_fuzzer.go:131]: WDYT?
```diff
diff --git a/tests/fuzzers/txfetcher/txfetcher_fuzzer.go b/tests/fuzzers/txfetcher/txfetcher_fuzzer.go
index 6c4526bca8..8b501645b6 100644
--- a/tests/fuzzers/txfetcher/txfetcher_fuzzer.go
+++ b/tests/fuzzers/txfetcher/txfetcher_fuzzer.go
@@ -117,6 +117,8 @@ func Fuzz(input []byte) int {
                        var (
                                announceIdxs = make([]int, announce)
                                announces    = make([]common.Hash, announce)
+                               types        = make([]byte, announce)
+                               sizes        = make([]uint32, announce)
                        )
                        for i := 0; i < len(announces); i++ {
                                annBuf := make([]byte, 2)
@@ -125,11 +127,13 @@ func Fuzz(input []byte) int {
                                }
                                announceIdxs[i] = (int(annBuf[0])*256 + int(annBuf[1])) % len(txs)
                                announces[i] = txs[announceIdxs[i]].Hash()
+                               types[i] = txs[announceIdxs[i]].Type()
+                               sizes[i] = uint32(txs[announceIdxs[i]].Size())
                        }
                        if verbose {
                                fmt.Println("Notify", peer, announceIdxs)
                        }
-                       if err := f.Notify(peer, nil, nil, announces); err != nil {
+                       if err := f.Notify(peer, types, sizes, announces); err != nil {
                                panic(err)
                        }
```

Comment (karalabe) [reply tests/fuzzers/txfetcher/txfetcher_fuzzer.go:131]: SGTM

File: eth/fetcher/tx_fetcher.go [hunk 2]
```diff
@@ -545,10 +577,27 @@ func (f *TxFetcher) loop() {
...
 				if _, ok := f.waitlist[hash]; ok {
 					for peer, txset := range f.waitslots {
+						if meta := txset[hash]; meta != nil {
+							if delivery.metas[i].kind != meta.kind {
+								log.Warn("Announced transaction type mismatch", "peer", peer, "tx", hash, "type", delivery.metas[i].kind, "ann", meta.kind)
+								f.dropPeer(peer)
+							} else if delivery.metas[i].size != meta.size {
+								log.Warn("Announced transaction size mismatch", "peer", peer, "tx", hash, "size", delivery.metas[i].size, "ann", meta.size)
+								if math.Abs(float64(delivery.metas[i].size)-float64(meta.size)) > 8 {
+									// Normally we should drop a peer considering this is a protocol violation.
```
Comment (holiman) [inline eth/fetcher/tx_fetcher.go:592]: Impressive level of indentation reached here. `switch case for if for if if if`. I don't have any great suggestions, but if you can make it less that would be supernice

Comment (holiman) [reply eth/fetcher/tx_fetcher.go:592]: Well, it's very hard to follow the code here. Like, you check `if _, ok := f.waitlist[hash]; ok {`, (but never use whatever object it is inside that waitlist, but if it's present, you check the waitslots, otherwise you operate on the announces. 

It might make perfect sense to you who wrote the code, but for a reviewer (and maybe even for yourself two years from now), it's pretty hard to ensure that it does what it should, and that all conditions are correct.

Comment (holiman) [reply eth/fetcher/tx_fetcher.go:592]: For example, couldn't it look something like this? 
```golang
f.onDeliveries(delivery, func(announced, actual *txMetadata){
				if announced.kind != actual.kind {
					log.Warn("Announced transaction type mismatch", "peer", peer, "tx", hash, "type", announced.kind, "ann", actual.kind)
					f.dropPeer(peer)
				} else if announced.size != meta.size {
					log.Warn("Announced transaction size mismatch", "peer", peer, "tx", hash, "size", announced.size, "ann", actual.size)
					if math.Abs(float64(announced.size)-float64(actual.size)) > 8 {
						// Normally we should drop a peer considering this is a protocol violation.
						// However, due to the RLP vs consensus format messyness, allow a few bytes
						// wiggle-room where we only warn, but don't drop.
						//
						// TODO(karalabe): Get rid of this relaxation when clients are proven stable.
						f.dropPeer(peer)
					}
				}})
```

Comment (MariusVanDerWijden) [reply eth/fetcher/tx_fetcher.go:592]: Since this logic is used twice here wouldn't it make sense to combine it into an anonymous function?
```
verifyMetadata := func(meta *txMetaData) {
    if meta == nil {
       return
    }
    if delivery.metas[i].kind != meta.kind {
    ...
```

Comment (MariusVanDerWijden) [reply eth/fetcher/tx_fetcher.go:592]: Ah damn, Martin was quicker :D

Comment (karalabe) [reply eth/fetcher/tx_fetcher.go:592]: You could probably clean the code up, but at this point I need to get this 5 liner change out of the way and focus on the rest of 4844 and don't want to diverge a month into splitting up the fetcher nice and elegantly. Though TBH, apart from introducing 10 more functions, I don't think it's gonna be much more elegant.

Comment (holiman) [reply eth/fetcher/tx_fetcher.go:592]: Sure, but I'll take that "You could probably clean the code up" literally, and after we merge this PR, I want to clean it up. For example like this:

https://github.com/karalabe/go-ethereum/compare/enforce-tx-announcement-metas...holiman:go-ethereum:enforce-tx-announcement-metas-mod?expand=1

Even if it means "introducing 10 more functions", I think 10 separate functions is preferrable over a gigantic switch with hundred-liner cases; it increases both readability and testability.

File: eth/fetcher/tx_fetcher.go [hunk 3]
```diff
@@ -422,22 +448,28 @@ func (f *TxFetcher) loop() {
...
 				// If the transaction is already known to the fetcher, but not
 				// yet downloading, add the peer as an alternate origin in the
 				// waiting list.
 				if f.waitlist[hash] != nil {
+					// Ignore double announcements from the same peer. This is
+					// especially important if metadata is also passed along to
+					// prevent malicious peers flip-flopping good/bad values.
+					if _, ok := f.waitlist[hash][ann.origin]; ok {
+						continue
+					}
```
Comment (holiman) [inline eth/fetcher/tx_fetcher.go:466]: Just want to double-check the logic here. So, say `tx A` has size `N`. Malicious peer `A` announces it to us, as size `M` and peer `B` announces it as size `N` (correctly).
Now, `peer A` does not deliver, so we instead ask peer `B` for it, and peer `B` delivers a tx with size `N`. 

- Do we hold both meta's, and thus expect `A` to have `M` if coming from `A` and `N` if coming from `B`? 
- If not, how do we prevent an attack where a malicious announcement is used to punish an honest tx-delivery?

Comment (holiman) [reply eth/fetcher/tx_fetcher.go:466]: And also, would it make sense to somehow flag (warn) if we notice this behaviour: peers announcing different meta for a transaction. 
It could point to shenanigans of this sort, and also of implementation flaws.

Comment (holiman) [reply eth/fetcher/tx_fetcher.go:466]: Answering my own question here: 

> * Do we hold both meta's, and thus expect A to have M if coming from A and N if coming from B?

Yes. So attack not possible here

Comment (holiman) [reply eth/fetcher/tx_fetcher.go:466]: ... and furthermore, once the honest peer delivers `tx A`, the lie from `peer A` will be discovered and it will be ejected, since we iterate over all announcements in `f.announces`. Good

Comment (karalabe) [reply eth/fetcher/tx_fetcher.go:466]: > And also, would it make sense to somehow flag (warn) if we notice this behaviour: peers announcing different meta for a transaction.
It could point to shenanigans of this sort, and also of implementation flaws.

In the end, it will turn into a Warn when the bad peer gets dropped. IMO there's no need to find announce clashes at the moment of announcements, when delivery happens we'll know for sure who's right and if anyone's wrong.

I guess what we could do is to add a second warning log where we log that someone was a "bit" off, but not off enough to drop.

File: eth/fetcher/tx_fetcher.go (hunk 2 above)
Comment (karalabe) [reply eth/fetcher/tx_fetcher.go:592]: :+1:

Comment (karalabe) [reply eth/fetcher/tx_fetcher.go:592]: The catch with a lot of small methods is that people have the tendency to reuse them. Ok, this sounds weird. But the current logic has a very very strict flow. When there are many methods, people tend to maybe turn one or the other into more generic ones, tweak them, reuse them and then the flow can get completely whacked out because someone forgot some ordering constaint. That's usually the reason I do large methods. It's hard to mess it up in the future.

**Review by holiman — APPROVED** [review]
Looks somewhat ok ish to me ...

**Comment by ngotchac** [comment]
Just noting that we are getting spammed (on the latest 1.13.3 release) with
```
Announced transaction size mismatch
```
warnings ; always with a delta between announced and actual size of 2 or 3 bytes. As explained in this PR, these warnings are for now ignored (for small deltas), until all implementations behave correctly. Shouldn't we also decrease the log severity?

PS : it seems that these warnings are all for some Erigon peers, if that helps...

**Comment by MariusVanDerWijden** [comment]
@ngotchac yep we're aware that erigon reports the sizes wrong. Thats why the PR allows for 8 bytes of difference between the announced and actual size. I've opened an issue here: https://github.com/ledgerwatch/erigon/issues/8456
